src/v/serde/envelope.h

```diff
@@ -67,77 +68,59 @@ struct checksum_envelope {
     static constexpr auto redpanda_serde_build_checksum = true;
 };
 
+// clang-format off
 namespace detail {
 
-template<typename T, typename = void>
-struct has_compat_attribute : std::false_type {};
-
 template<typename T>
-struct has_compat_attribute<
-  T,
-  std::void_t<decltype(std::declval<T>().redpanda_serde_compat_version)>>
-  : std::true_type {};
-
-template<typename T, typename = void>
-struct has_version_attribute : std::false_type {};
-
-template<typename T>
-struct has_version_attribute<
-  T,
-  std::void_t<decltype(std::declval<T>().redpanda_serde_version)>>
-  : std::true_type {};
-
-template<typename T, typename = void>
-struct inherits_from_envelope : std::false_type {};
+concept has_compat_attribute = requires(T t) {
+    t.redpanda_serde_compat_version;
+};
```

Comment:
there is a lot going on in this one commit. might be easier to review if it is split up

Reply:
sure, i tried to split them up. hope it helps.